File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/structure/Entity.java
package net.nemerosa.ontrack.model.structure;

import static org.apache.commons.lang3.Validate.isTrue;

/**
 * An <b>Entity</b> is a model object that has an {@link ID}. The state of this ID will determined the status
 * of this entity:
 * <ul>
 * <li><i>new</i> - the ID is not {@linkplain ID#isSet() set}.</li>
 * <li><i>defined</i> - the ID is {@linkplain ID#isSet() set}.</li>
 * </ul>
 */
public interface Entity {

    public static void isEntityNew(Entity e, String message) {
        isTrue(e != null && !ID.isDefined(e.getId()), message);
    }

    public static void isEntityDefined(Entity e, String message) {
        isTrue(e != null && ID.isDefined(e.getId()), message);
    }

    ID getId();

    default int id() {
        ID id = getId();
        isTrue(ID.isDefined(id), "ID must be defined");
        return getId().getValue();
    }

}

File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/structure/Entity.kt
package net.nemerosa.ontrack.model.structure

import net.nemerosa.ontrack.model.structure.ID.Companion.isDefined

/**
 * An `Entity` is a model object that has an [ID]. The state of this ID will determined the status
 * of this entity:
 *
 *  * *new* - the ID is not [set][ID.isSet].
 *  * *defined* - the ID is [set][ID.isSet].
 *
 */
interface Entity {

    /**
     * ID of the entity
     */
    val id: ID

    /**
     * Numeric value of the ID
     *
     * @throws IllegalStateException If the entity is not defined.
     */
    fun id(): Int {
        check(isDefined(id)) {
            "ID of the entity must be defined"
        }
        return id.value
    }

    companion object {

        @JvmStatic
        fun isEntityNew(e: Entity?, message: String) {
            check(e != null && !isDefined(e.id)) { message }
        }

        @JvmStatic
        fun isEntityDefined(e: Entity?, message: String) {
            check(e != null && isDefined(e.id)) { message }
        }
    }
}